usb: device_next: Add Kconfig to set maximum speed

tmon-nordic opened the pull request:
Setting maximum supported speed to Full-Speed can reduce code size by allowing compiler optimizations at compile time and also helps reducing RAM by allocating smaller bulk buffers. More RAM optimizations are possible but this commit only gets the low hanging fruits.

Another idea is that High-Speed capable drivers can use this information to limit operating speed to user choice. This PR implements this only for udc_dwc2 driver.

Implements: #75002

tmon-nordic committed:
usb: device_next: Unregister class according to speed

Fix a typo resulted in USBD_SPEED_HS unregistering Full-Speed class
instead of High-Speed.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: Add Kconfig to set maximum speed

Two main ideas behind setting maximum speed are:
  * Allow code and RAM optimizations at compile time
  * Allow High-Speed capable drivers to limit operating speed to user
    choice.

This commit only introduces the necessary Kconfig options but does not
implement any code or RAM optimizations and does not modify any driver.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

henrikbrixandersen commented on subsys/usb/device_next/Kconfig:
```diff
@@ -16,6 +16,24 @@ module = USBD
+
+endchoice
+
+config USBD_MAX_SPEED
+	int
+	default 0 if USBD_MAX_SPEED_FULL
+	default 1 if USBD_MAX_SPEED_HIGH
```
How come this indirection is needed? Why not simple do `if (IS_ENABLED(CONFIG_USBD_MAX_SPEED_HIGH)` in the C code?

tmon-nordic replied:
The choice here makes it possible to add Super-Speed support later on without much hassle. `IS_ENABLED(CONFIG_USBD_MAX_SPEED_HIGH)` would make things harder IMHO.

henrikbrixandersen replied:
That makes sense, thanks. I'll take a closer look later today.

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -1848,14 +1849,14 @@ static int udc_dwc2_init_controller(const struct device *dev)
-			<< USB_DWC2_DCFG_DEVSPD_POS;
+		dcfg |= usb_dwc2_set_dcfg_devspd(USBD_SUPPORTS_HIGH_SPEED ?
+		    USB_DWC2_DCFG_DEVSPD_USBHS20 : USB_DWC2_DCFG_DEVSPD_USBFS20);
 		break;
 	case USB_DWC2_GHWCFG2_HSPHYTYPE_UTMIPLUS:
 		gusbcfg |= USB_DWC2_GUSBCFG_PHYSEL_USB20 |
 			   USB_DWC2_GUSBCFG_ULPI_UTMI_SEL_UTMI;
-		dcfg |= USB_DWC2_DCFG_DEVSPD_USBHS20
-			<< USB_DWC2_DCFG_DEVSPD_POS;
+		dcfg |= usb_dwc2_set_dcfg_devspd(USBD_SUPPORTS_HIGH_SPEED ?
+		    USB_DWC2_DCFG_DEVSPD_USBHS20 : USB_DWC2_DCFG_DEVSPD_USBFS20);
```
This layer violation. You may not use anything from subsys/usb/device_next. And there is absolutely no need to limit the high-speed controller to full speed.

tmon-nordic replied:
There are valid reasons for disabling high-speed, otherwise the peripherals wouldn't support it.

tmon-nordic replied:
Setting the reasons aside (the most trivial one is replicating an existing device with a newer SoC in a way that allows reusing existing host software without any modifications), please guide how to achieve this functionality without "layer violation". Note that USB itself has design choices that could be classified as layering violations, like:
* different configurations based on operating speed
* the meaning of bInterval field in Endpoint descriptor depends on operating speed
* isochronous feedback size and meaning depends on operating speed

The operating speed plays so significant role in USB that I believe there must be a way to force Full-Speed only operation of High-Speed capable device without resorting to filtering out high-speed detection handshake on D+/D- lines.

jfischer-no commented on subsys/usb/device_next/Kconfig:
```diff
@@ -16,6 +16,24 @@ module = USBD
 module-str = usbd
 source "subsys/logging/Kconfig.template.log_config"
 
+choice USBD_MAX_SPEED_CHOICE
+	prompt "Max supported connection speed"
+	default USBD_MAX_SPEED_FULL if UDC_IT82XX2 || UDC_KINETIS || UDC_NRF || UDC_STM32
+	default USBD_MAX_SPEED_HIGH
```
This layer violation. You may not use anything from drivers/usb/*/Kconfig.* or have any dependencies on drivers.

tmon-nordic replied:
Then how do you enable the easy and quite significant code and RAM savings on platforms that have controllers that are not capable of high-speed operation? Please come up with alternative instead of just screaming "layer violation".

tmon-nordic replied:
Let's consider CDC ACM sample to be representative one. When building it for nRF52840DK (`CONF_FILE=usbd_next_prj.conf west build --pristine -b nrf52840dk/nrf52840 -d build-cdc-acm-next ~/zephyrproject/zephyr/samples/subsys/usb/cdc_acm`) on 596ef0a519eec584e2c403927543b130a7166a45 the summary is:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       61636 B         1 MB      5.88%
             RAM:       20976 B       256 KB      8.00%
        IDT_LIST:          0 GB        32 KB      0.00%
```
With this PR the summary is:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       60924 B         1 MB      5.81%
             RAM:       20080 B       256 KB      7.66%
        IDT_LIST:          0 GB        32 KB      0.00%
```

Note that this PR does only the low-hanging fruits and more savings would be possible (but would require a lot of `COND_CODE` which may be too costly maintenance-wise).

jfischer-no commented on subsys/usb/device_next/Kconfig:
```diff
@@ -16,6 +16,23 @@ module = USBD
 source "subsys/logging/Kconfig.template.log_config"
 
+choice USBD_MAX_SPEED_CHOICE
+	prompt "Max supported connection speed"
```
`default USBD_MAX_SPEED_HIGH if UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT`

bjarki-andreasen reviewed (dismissed):
The speed choice looks good now :)

tmon-nordic commented:
Rebased to fix merge conflicts.

tmon-nordic committed:
drivers: udc_dwc2: Limit operating speed to stack configuration

Limit maximum operating speed in DCFG register if USB stack is
configured to support only Full-Speed.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic commented:
Rebased to fix merge conflicts.

jfischer-no commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -138,7 +138,8 @@ static uint8_t bt_hci_get_bulk_in(struct usbd_class_data *const c_data)
 	struct usbd_bt_hci_desc *desc = data->desc;
 
-	if (usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
+	if (USBD_SUPPORTS_HIGH_SPEED &&
+	    usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
```
This does not look good and is easy to forget. I suggest adding a helper macro that could take usbd_bus_speed(), usbd_caps_speed_is_hs() or just speed as argument.

```diff
diff --git a/include/zephyr/usb/usbd.h b/include/zephyr/usb/usbd.h
index 186f79b991b..8ff45bfe678 100644
--- a/include/zephyr/usb/usbd.h
+++ b/include/zephyr/usb/usbd.h
@@ -966,6 +966,15 @@ int usbd_wakeup_request(struct usbd_context *uds_ctx);
  */
 enum usbd_speed usbd_bus_speed(const struct usbd_context *const uds_ctx);
 
+/**
+ * @brief Helper to check if speed is high speed
+ *
+ * @param[in] fspeed Parameter that expands to usbd_speed
+ *
+ * @return 1 if actual speed is high speed
+ */
+#define usbd_speed_is_hs(fspeed) \
+       (USBD_SUPPORTS_HIGH_SPEED && fspeed == USBD_SPEED_HS)

 /**
  * @brief Get highest speed supported by the controller
  *
```

Maybe also shorthand for usbd_bus_speed()/usbd_caps_speed()

```c
+/**
+ * @brief Helper to check if actual device speed is high speed
+ *
+ * @param[in] uds_ctx Pointer to a device context
+ *
+ * @return 1 if actual device speed is high speed
+ */
+#define usbd_bus_speed_is_hs(uds_ctx) \
+       (usbd_speed_is_hs(usbd_bus_speed(uds_ctx)))
```

tmon-nordic replied:
I tend to disagree. Hiding compile time evaluation conditions behind a macro that looks just like a function is IMHO hiding the fact that the code is supposed to go away when compiled with Full-Speed only configuration.

I won't object the macros if you submit the PR yourself, but I just don't think these are justified myself.

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -18,6 +18,7 @@
 #include <zephyr/sys/byteorder.h>
 #include <zephyr/drivers/usb/udc.h>
 #include <zephyr/usb/usb_ch9.h>
+#include <zephyr/usb/usbd.h>
```
You must not include a stack header in the driver, this is mixing different layers. This will never be approved by me. I can show you another way to force a driver to run at full speed if you address my first comment.

jfischer-no replied:
I reworked your approach and pushed it to https://github.com/jfischer-no/zephyr/commits/pr-76255/. With https://github.com/jfischer-no/zephyr/commit/a9335f66218b94bb19d013a9c096c44052890aa0, the settings in the UDC driver are independent and do not interfere with stack. I also changed stack's Kconfig in https://github.com/jfischer-no/zephyr/commit/df95264e6a50158c922392cbdd38ac8307471226, see the commit message or better try it out.

tmon-nordic replied:
I did try out your changes and it is possible to build High-Speed capable USB device stack with Full-Speed only driver. Such configuration doesn't make sense to me. What's the point? Why make it so complex?

jfischer-no replied:
It is no more complicated than the original. It always defaults to a valid configuration. The high-speed part always compiles. But yes, it allows you to build a stack that has high-speed support even if the device does not. This is intentional, for example to see the difference in flash/RAM usage.

tmon-nordic replied:
I think that intention is pretty weird. @carlescufi can you please weigh in?

tmon-nordic committed:
usb: device_next: Reduce code size on Full-Speed only configuration

Allow compiler optimizations to remove High-Speed handling code. Knowing
that maximum operating speed is Full-Speed allows to reduce bulk buffers
from 512 to 64 bytes. More RAM optimizations are possible but this
commit only gets the low hanging fruits.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic commented on drivers/usb/udc/Kconfig:
```diff
@@ -11,6 +11,9 @@ menuconfig UDC_DRIVER
 
 if UDC_DRIVER
 
+config UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT
```
Done.

bjarki-andreasen commented:
Given the usb controller drivers are hidden behind the `udc.h` API, I believe the USB drivers should not depend on any config from the USBD stack.

For RTC drivers, we have configs like `CONFIG_RTC_ALARM`, this config belongs to the RTC driver subsystem, and any higher layer must select it or depend on it. The RTC drivers do not depend on any higher layer config. I agree with @jfischer-no here.

Since the USBD stack is higher layer, it has to depend on a config like `UDC_DRIVER_HIGH_SPEED_SUPPORT`, note I inverted the config to be a "whitelist", which a user or subsystem needs to select. It should depend on `UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT` which is selected by drivers so its not selected with no implementation.

As a user, I want the USBD stack supporting some speed, so I want to do
```
CONFIG_USBD_MAX_SPEED_HIGH=y
```
and be done with it. USBD then tries to select everything it needs from the lower layer:
```
config USBD_MAX_SPEED_HIGH
        select UDC_DRIVER_HIGH_SPEED_SUPPORT
        select UDC_DRIVER
```
and it will fail if no UDC supports the requested max speed.

I know this is a third option, sorry for that, but I think it solves the issue without violating the driver layer as pointed out by @jfischer-no

tmon-nordic commented:
> As a user, I want the USBD stack supporting some speed, so I want to do
> 
> ```
> CONFIG_USBD_MAX_SPEED_HIGH=y
> ```
> 
> and be done with it. USBD then tries to select everything it needs from the lower layer:
> 
> ```
> config USBD_MAX_SPEED_HIGH
>         select UDC_DRIVER_HIGH_SPEED_SUPPORT
>         select UDC_DRIVER
> ```
> 
> and it will fail if no UDC supports the requested max speed.
> 
> I know this is a third option, sorry for that, but I think it solves the issue without violating the driver layer as pointed out by @jfischer-no

This is discussed and considered to be the equivalent for Option 1. This just shifts the indirectness of the `USBD_MAX_SPEED_HIGH` dependency from a define in header file to Kconfig (i.e. Kconfig engine would be doing the thing instead of code).

I am in favor of either the Option 1 or Option 3 because they eliminate the unintended possibility for configs like "bloated high-speed stack with full-speed only controller".

tmon-nordic commented:
> It (theoretically) would allow for different speed limits to be set per driver in case of multiple USB device drivers being available on the same system.

No, it wouldn't because the option is global for all UDC drivers and not per-driver.

jfischer-no commented:
> As a user, I want the USBD stack supporting some speed, so I want to do
> 
> ```
> CONFIG_USBD_MAX_SPEED_HIGH=y
> ```
> 
> and be done with it. USBD then tries to select everything it needs from the lower layer:
> 
> ```
> config USBD_MAX_SPEED_HIGH
>         select UDC_DRIVER_HIGH_SPEED_SUPPORT
>         select UDC_DRIVER
> ```
> 
> and it will fail if no UDC supports the requested max speed.
> 
> I know this is a third option, sorry for that, but I think it solves the issue without violating the driver layer as pointed out by @jfischer-no

@bjarki-andreasen No problem, thanks for proposal, but I insist that information about controller capabilities should be provided and set at the driver level.

In general, I am confident that it should not be possible to limit the speed of the controller. There is no benefit to doing so in the real life of a USB device, and we should not bloat Kconfig with unnecessary options. However if people still want to do it: **there must be no dependency in the driver on the higher level, nor should higher level change anything in the driver other than to enable it**. This is fundamental, with no exceptions.

The information about the controller capability is provided by driver through the driver API. Some platforms use DT to describe supported speed, the other can get this information form the controller hardware at runtime. For example, there could be two controllers on a SoC, the first supports high speed and the second only supports only full speed. The stack will get this information at the runtime (also the actual speed) using UDC API, and handle the device accordingly. The DT must not be used at a higher level (and there is no need to do it). If we want to optimize the stack and do not enable high-speed device-specific handling by default if the used controller does not support it, then the Kconfig option `UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT` is the way to go. If the user want to force high-speed capable controller to full speed operation only, then it must be forced at the driver level to be consistent source of information and not introduce the dependency on higher level software. This was provided by me in the branch referred in the option 2. New Option 3 does not fulfill the above requirements.

bjarki-andreasen commented:
> If we want to optimize the stack and do not enable high-speed device-specific handling by default if the used controller does not support it, then the Kconfig option UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT is the way to go. If the user want to force high-speed capable controller to full speed operation only, then it must be forced at the driver level to be consistent source of information and not introduce the dependency on higher level software. This was provided by me in the branch referred in the option 2. New Option 3 does not fulfill the above requirements.

As I understand it, like `RTC`, `UDC` is the driver level, the drivers implement the `struct udc_api`. The `config UDC_DRIVER_` configs are driver level configs. Do you consider the `struct udc_api` a higher level than the driver? or am I misunderstanding something?

tmon-nordic commented:
> # Disable the USB2 communication, i.e. only USB3 wires going to a laptop lid webcam

This won't be standard compliant and you won't be able to certify such solution.

> I would like to ask, why not just depend on devicetree for this in option 2, though? The user could set the maximum supported speed for their application/board per USB device controller and have the driver Kconfig pick this up?

devicetree is to mostly to describe hardware capabilities. While Zephyr does use it more for configuration, I think the UDC nodes should really describe what the hardware is capable of.

> I may be misunderstanding what the driver level config is meant to do, from @jfischer-no comment, is there no code size benefit to limit the driver to a lower speed? like disabling the RTC alarm implementation?

I don't think that you can get any real code size benefits at the UDC driver level by limiting the speed. You can get significant code size and RAM benefits for removing high-speed support (or any higher speeds) at the USB stack level.

> In general, I am confident that it should not be possible to limit the speed of the controller. In general, I am confident that it should not be possible to limit the speed of the controller. There is no benefit to doing so in the real life of a USB device

Yes, there are real use cases for doing so. One example would be if the user wants to implement Basic Audio Device. This is only allowed on Full-Speed only devices (it is prohibited at High-Speed by the specification). Are you suggesting that the user must not choose SoC with High-Speed capable controller just because they want to use it in Full-Speed only application? There are way more factors involved when choosing components than the maximum supported USB speed (and especially that every single High-Speed capable controller is guaranteed to be able to operate at Full-Speed, otherwise it wouldn't be possible to certify such controller).

Another examples are the many cases where some product has EOL components and there is a business need for a new production run - the real world in such cases looks just to replicate the existing functionality in just the part (device) that needs replacement. There are real benefits if for every other part in the system the replacement is transparent (products do change components used on the PCB level while retaining exactly the same packaging, model information, etc.).

> and we should not bloat Kconfig with unnecessary options 

This I completely agree. And I do see the UDC Kconfig options for limiting the speed as the bloat. If you want to not have unnecessary options, then there must be only one place to select the maximum operating speed.

> If the user want to force high-speed capable controller to full speed operation only, then it must be forced at the driver level to be consistent source of information and not introduce the dependency on higher level software

But then you have the layer violation in the USB stack if the only source of information is to be at the driver level.

bjarki-andreasen commented:
What about passing the maximum speed, or enabling specific speeds, using a runtime API like
```
int udc_device_set_supported_speed(const struct device *dev,
                                   bool support_full_speed,
                                   bool support_high_speed,
                                   bool support_super_speed);
```
That allows the higher layers to tell the udc drivers which speeds to accept, based on what is implemented in said higher layer. It could even be part of `udc_init()` maybe? No driver level kconfig needed :)

tmon-nordic commented:
> What about passing the maximum speed, or enabling specific speeds, using a runtime API

This would actually slightly increase the size of drivers that are capable of more than just Full-Speed. I think the hidden Kconfig symbols like in Option 3 are a good tradeoff between limiting the "layer crossing" and allowing compile-time optimizations.

Note that we currently do not support Low-Speed, but if at later point there is some target with Low-Speed only (or Low-Speed as an option and someone wants to use it) then the Option 1 and Option 3 can easily be extended to cover the use case. The idea is to have just a single `choice` item in menuconfig related to setting the speed. Low-Speed only selection would actually prevent the classes that use bulk and/or isochronous transfers, e.g. CDC ACM, MSC, UAC2.

carlescufi commented:
> In general, I am confident that it should not be possible to limit the speed of the controller. There is no benefit to doing so in the real life of a USB device, and we should not bloat Kconfig with unnecessary options. However if people still want to do it: **there must be no dependency in the driver on the higher level, nor should higher level change anything in the driver other than to enable it**. This is fundamental, with no exceptions.

But I don't get it, with [Option 3](https://github.com/zephyrproject-rtos/zephyr/pull/76255#issuecomment-2682562086) you don't have a dependency in the driver on the higher level? and the higher level is not changing anything other than enabling it at a particular speed?

tmon-nordic commented:
This PR achieves two main goals:
  1. Reduce code and RAM usage on Full-Speed only targets
  2. Allow developer to limit the maximum operating speed
  
The design that enables above goals boils down to:
  * Every driver that supports High-Speed selects `UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT`
  * There is one and only one user-configurable Kconfig `USBD_MAX_SPEED_CHOICE`
    - the choice defaults to maximum supported speed (High-Speed if selected)
    - High-Speed choice is only available when `UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT` is selected
    - user can always lower the speed (USB mandates that every High-Speed device must allow Full-Speed operation)
    - If and only if the stack is configured to support High-Speed, the hidden Kconfig `UDC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED` is selected
  * When Full-Speed is the selected maximum speed, then High-Speed support is gone from stack (at compile time)
    - USB classes use `if (USBD_SUPPORTS_HIGH_SPEED)` construct to wrap High-Speed only functionality
  * High-Speed capable drivers, configure peripheral to limit maximum operating speed based on hidden Kconfig `UDC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED`

Architecture Working Group is expected to resolve following disagreement:

**a) Whether allowing user to limit High-Speed capable controller to operate at Full-Speed only is supported use case.**
- @tmon-nordic strongly believes this is normal use case, because USB controller is just one of many features that are determining the SoC selection. Every High-Speed capable controller can operate at Full-Speed (otherwise it wouldn't be certified), therefore the user can search for SoCs with Full-Speed or High-Speed capable controllers if the end application will only ever operate at Full-Speed.

**b) Whether having a single user-configurable source of truth for maximum supported speed (Kconfig choice `USBD_MAX_SPEED_CHOICE`) is appropriate**

---

Future extensibility considerations:
  * Super-Speed devices with USB connector must support at least one USB 2.x speed (Low/Full/High)
    - currently USB stack does not support Low speed at all. Low-Speed support can be added on base of this PR later on without affecting overall user experience.
    - currently Full-Speed is always enabled
    - the Super-Speed devices could have separate choice entry for:
      Super-Speed + High-Speed (Full-Speed support is implied by USB specification)
      Super-Speed + Full-Speed
      Super-Speed + Low-Speed (when stack supports Low-Speed)
  * The additional choices (entries in Kconfig selection box) could be used to determine:
    - Macros values: `USBD_SUPPORTS_HIGH_SPEED`, `USBD_SUPPORTS_SUPER_SPEED` (and possibly `USBD_SUPPORTS_FULL_SPEED` and `USBD_SUPPORTS_LOW_SPEED`)
    - UDC hidden Kconfig symbols: `UDC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED`, `UDC_DRIVER_SUPER_SPEED_SUPPORT_ENABLED` (and possibly `UDC_DRIVER_FULL_SPEED_SUPPORT_ENABLED`, `UDC_DRIVER_LOW_SPEED_SUPPORT_ENABLED`)

henrikbrixandersen commented:
> The Option 3 is now implemented in this PR.

For completeness (and to avoid confusion), could we have option 3 added to the table in https://github.com/zephyrproject-rtos/zephyr/pull/76255#issuecomment-2682051627 (or create a new table with all three options listed)?

carlescufi commented:
A quick note from me as part of the upcoming discussion:

@jfischer-no wrote:
> In general, I am confident that it should not be possible to limit the speed of the controller. There is no benefit to doing so in the real life of a USB device, and we should not bloat Kconfig with unnecessary options

Not only there is a tangible benefit, but Nordic has specific requests from customers to enable this. The reasoning is simple, as outlined by the customer: smaller code size, reduced complexity, ability to reuse parts of verification/internal certification.

jfischer-no commented:
> > In general, I am confident that it should not be possible to limit the speed of the controller. There is no benefit to doing so in the real life of a USB device, and we should not bloat Kconfig with unnecessary options
> 
> Not only there is a tangible benefit, but Nordic has specific requests from customers to implement this. The reasoning is simple, as outlined by the customer: smaller code size, reduced complexity, ability to reuse parts of verification/internal certification.

This would only reduce memory usage. It does not reduce code size, the code is already there and is always built. This PR actually increases complexity. Again, for a USB device, if you replace a full-speed device with another that is a high-speed device, then no verification/certification can be reused because it is a new device regardless of speed.

cfriedt commented:
> > I would like to ask, why not just depend on devicetree for this in option 2, though? The user could set the maximum supported speed for their application/board per USB device controller and have the driver Kconfig pick this up?
> 
> devicetree is to mostly to describe hardware capabilities. While Zephyr does use it more for configuration, I think the UDC nodes should really describe what the hardware is capable of.

That was the question that immediately jumped out to me as well.

decsny commented:
@jfischer-no is there some reason that the driver layer cant have a choice-type config structure to get the "ease of use" like in option 1/3 but not depend on upper layer?

decsny commented:
> > > I would like to ask, why not just depend on devicetree for this in option 2, though? The user could set the maximum supported speed for their application/board per USB device controller and have the driver Kconfig pick this up?
> > 
> > 
> > devicetree is to mostly to describe hardware capabilities. While Zephyr does use it more for configuration, I think the UDC nodes should really describe what the hardware is capable of.
> 
> That was the question that immediately jumped out to me as well.

yet another case of zephyr missing a multi instance static software config system

jfischer-no commented:
> @jfischer-no is there some reason that the driver layer cant have a choice-type config structure to get the "ease of use" like in option 1/3 but not depend on upper layer?

See https://github.com/zephyrproject-rtos/zephyr/pull/76255#issuecomment-2685164580
Also, option 3 does not provide the solution to problem "allowing user to limit High-Speed capable controller to operate at Full-Speed only is supported use case" and "having a single user-configurable source of truth for maximum supported speed". Only option 2 is consistently correct, because if you want to limit the controller, you should do it at the controller driver level, and the controller driver already is and will remain the "single source of truth for maximum supported speed".

jfischer-no commented:
> > > > I would like to ask, why not just depend on devicetree for this in option 2, though? The user could set the maximum supported speed for their application/board per USB device controller and have the driver Kconfig pick this up?
> > > 
> > > 
> > > devicetree is to mostly to describe hardware capabilities. While Zephyr does use it more for configuration, I think the UDC nodes should really describe what the hardware is capable of.
> > 
> > 
> > That was the question that immediately jumped out to me as well.
> 
> yet another case of zephyr missing a multi instance static software config system

This PR has nothing to do with DT or multi-instance static software configuration system. There is no dependency on DT in the USB device stack next.

tmon-nordic commented:
> Also, option 3 does not provide the solution to problem "allowing user to limit High-Speed capable controller to operate at Full-Speed only is supported use case" and "having a single user-configurable source of truth for maximum supported speed".

I disagree. Option 1/3 is precisely what has the one and only one Kconfig option to limit the application as a whole to Full-Speed only.

> Only option 2 is consistently correct, because if you want to limit the controller, you should do it at the controller driver level, and the controller driver already is and will remain the "single source of truth for maximum supported speed".

You don't seem to accept that the real world use case for the user is to limit the USB stack (external device behavior) for particular speed. Limiting the controller to disable High-Speed chirp is just a necessary consequence of the user choice to limit the stack.

decsny commented:
> > > > > I would like to ask, why not just depend on devicetree for this in option 2, though? The user could set the maximum supported speed for their application/board per USB device controller and have the driver Kconfig pick this up?
> > > > 
> > > > 
> > > > devicetree is to mostly to describe hardware capabilities. While Zephyr does use it more for configuration, I think the UDC nodes should really describe what the hardware is capable of.
> > > 
> > > 
> > > That was the question that immediately jumped out to me as well.
> > 
> > 
> > yet another case of zephyr missing a multi instance static software config system
> 
> This PR has nothing to do with DT or multi-instance static software configuration system. There is no dependency on DT in the USB device stack next.

yes I know, but someone else brought it up, and I think you can't deny it would be nice to be able configure the limit per device since many platforms have multiple devices (not endpoints, devices) that use the same driver. But it is outside of the scope of this PR so we don't need to discuss about it here, you're right



> > @jfischer-no is there some reason that the driver layer cant have a choice-type config structure to get the "ease of use" like in option 1/3 but not depend on upper layer?
> 
> See [#76255 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/76255#issuecomment-2685164580) Also, option 3 does not provide the solution to problem "allowing user to limit High-Speed capable controller to operate at Full-Speed only is supported use case" and "having a single user-configurable source of truth for maximum supported speed". Only option 2 is consistently correct, because if you want to limit the controller, you should do it at the controller driver level, and the controller driver already is and will remain the "single source of truth for maximum supported speed".

I read your comment and you said

> In general, I am confident that it should not be possible to limit the speed of the controller. There is no benefit to doing so in the real life of a USB device, and we should not bloat Kconfig with unnecessary options. 

I think we are past this by now in the discussion because it was made clear that the goal is to configure a device to act only as full speed for the purpose of footprint savings, whether you personally want to do this is besides the point. And as far as Kconfig bloat, I don't think that is even valid here.  I am willing to completely agree with you on this statement you made:

> there must be no dependency in the driver on the higher level, nor should higher level change anything in the driver other than to enable it. This is fundamental, with no exceptions.

because to me that seems like the most sound architecture, and therefore your option 2 seems preferable.  But I also have to say that at first looking at the comparison comment above , at a glance, the option 1/3 looks more preferable from a user perspective. Whether it is long term a good design is more important of course, but what I don't understand is why would you be opposed to a solution that combines your architectural awareness with the usability of the other two options.  It seems obvious to me here unless I don't understand something that we can get the best of both worlds simply by introducing the user choice config similar to what is option 1/3 at the driver level to be fundamentally like option 2.  What is the complication here?

tmon-nordic commented:
> it would be nice to be able configure the limit per device since many platforms have multiple devices (not endpoints, devices) that use the same driver. But it is outside of the scope of this PR so we don't need to discuss about it here, you're right

If the user chooses to limit USB functionality (to Full-Speed only) at the stack level, then it doesn't matter what underlying drivers are there (all drivers should be aware that they are operating with Full-Speed only stack and therefore should disable High-Speed chirp).

>  It seems obvious to me here unless I don't understand something that we can get the best of both worlds simply by introducing the user choice config similar to what is option 1/3 at the driver level to be fundamentally like option 2. What is the complication here, if there is any, from a technical (non opinionated) perspective? and don't say "bloat" or "unnecessary"

The technical point is that by having the user-selectable option in UDC to limit the speed, then we would have to either:
  * Ditch the USB stack limiting option completely, and base the stack limits based on UDC Kconfig option. OR
  * Allow the "Kconfig trap" that makes it possible to build "full-speed only driver with high-speed capable stack".

tmon-nordic committed:
usb: device_next: Change speed selection Kconfig dependency

In my opinion, the user is supposed to configure the speed of the stack
and drivers ough to honor that choice. However current Zephyr USB
maintainer imposes that the dependency is the other way round, i.e.
that user first needs to disable High-Speed chirp at driver level and
only then can select Full-Speed only operation. Adhere to the
arbitrarily set up rule to allow this really necessary functionality to
enter Zephyr.

I consider this change to be harmful because it opens up a Kconfig trap
that allows configuring High-Speed capable stack with a device driver
limited to Full-Speed only operation.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic commented:
This whole disagreement is summed up in the latest commit "usb: device_next: Change speed selection Kconfig dependency". I believe this commit is extremely harmful, but it may be the only way to move this forward.
